python/dgl/graphbolt/impl/neighbor_sampler.py

```diff
@@ -1,18 +1,125 @@
 """Neighbor subgraph samplers for GraphBolt."""
 
+from concurrent.futures import ThreadPoolExecutor
 from functools import partial
 
 import torch
 from torch.utils.data import functional_datapipe
+from torchdata.datapipes.iter import Mapper
 
 from ..internal import compact_csc_format, unique_and_compact_csc_formats
 from ..minibatch_transformer import MiniBatchTransformer
 
 from ..subgraph_sampler import SubgraphSampler
+from .fused_csc_sampling_graph import fused_csc_sampling_graph
 from .sampled_subgraph_impl import SampledSubgraphImpl
 
 
-__all__ = ["NeighborSampler", "LayerNeighborSampler"]
+__all__ = ["NeighborSampler", "LayerNeighborSampler", "SamplePerLayer"]
+
+
+@functional_datapipe("fetch_insubgraph_data")
+class FetchInsubgraphData(Mapper):
+    """Fetches the insubgraph and wraps it in a FusedCSCSamplingGraph object."""
+
+    def __init__(
+        self, datapipe, sample_per_layer_obj, stream=None, executor=None
+    ):
+        super().__init__(datapipe, self._fetch_per_layer)
+        self.graph = sample_per_layer_obj.sampler.__self__
+        self.prob_name = sample_per_layer_obj.prob_name
+        self.stream = stream
+        if executor is None:
+            self.executor = ThreadPoolExecutor(max_workers=1)
+        else:
+            self.executor = executor
+
+    def _fetch_per_layer_helper(self, minibatch, stream):
+        with torch.cuda.stream(self.stream):
+            index = minibatch._seed_nodes
+            if index.is_cuda:
+                index.record_stream(torch.cuda.current_stream())
+            index_select = partial(
+                torch.ops.graphbolt.index_select_csc, self.graph.csc_indptr
+            )
+
+            def record_stream(tensor):
+                if stream is not None and tensor.is_cuda:
+                    tensor.record_stream(stream)
+
+            indptr, indices = index_select(self.graph.indices, index, None)
+            record_stream(indptr)
+            record_stream(indices)
+            output_size = len(indices)
+            if self.graph.type_per_edge is not None:
+                _, type_per_edge = index_select(
+                    self.graph.type_per_edge, index, output_size
```

Comment:
Is this correct to use index or indices?

Reply:
Index because we defined `index_select` above as:
```python
            index_select = partial(
                torch.ops.graphbolt.index_select_csc, self.graph.csc_indptr
            )
```

Reply:
We are getting the vaue for the insubgraph neighbors of seed nodes when used like this. Do you think I can make this more clear in the code?

Reply:
I see, index_select is a reserved name for the specific op, reuse it here causes confusion. Consider rename it to index_select_csc_with_indptr or better ones.